pkg/cmd/search/code/code.go

```diff
@@ -0,0 +1,218 @@
+			$ gh search code cli --owner=microsoft
+
+			# search code matching "html" in repositories owned by octocat
+			$ gh search code html --owner=octocat
+
+			# search code matching "panic" in cli repository
+			$ gh search code panic --repo cli/cli
+
+			# search code matching keyword "lint" in package.json files
+			$ gh search code lint --filename package.json
+		`),
+		RunE: func(c *cobra.Command, args []string) error {
+			if len(args) == 0 && c.Flags().NFlag() == 0 {
+				return cmdutil.FlagErrorf("specify search keywords or flags")
+			}
+			if opts.Query.Limit < 1 || opts.Query.Limit > shared.SearchMaxResults {
+				return cmdutil.FlagErrorf("`--limit` must be between 1 and 1000")
+			}
+			opts.Query.Keywords = args
+			if runF != nil {
+				return runF(opts)
+			}
+			var err error
+			opts.Searcher, err = shared.Searcher(f)
+			if err != nil {
+				return err
+			}
+			return codeRun(opts)
+		},
+	}
+
+	// Output flags
+	cmdutil.AddJSONFlags(cmd, &opts.Exporter, search.CodeFields)
+	cmd.Flags().BoolVarP(&opts.WebMode, "web", "w", false, "Open the search query in the web browser")
+
+	// Query parameter flags
+	cmd.Flags().IntVarP(&opts.Query.Limit, "limit", "L", 30, "Maximum number of code results to fetch")
+
+	// Query qualifier flags
+	cmd.Flags().StringVar(&opts.Query.Qualifiers.Extension, "extension", "", "Filter on file extension")
+	cmd.Flags().StringVar(&opts.Query.Qualifiers.Filename, "filename", "", "Filter on filename")
+	cmdutil.StringSliceEnumFlag(cmd, &opts.Query.Qualifiers.In, "match", "", nil, []string{"file", "path"}, "Restrict search to file contents or file path")
+	cmd.Flags().StringVarP(&opts.Query.Qualifiers.Language, "language", "l", "", "Filter results by language")
+	cmd.Flags().StringSliceVar(&opts.Query.Qualifiers.Repo, "repo", nil, "Filter on repository")
+	cmd.Flags().StringVar(&opts.Query.Qualifiers.Size, "size", "", "Filter on size range, in kilobytes")
+	cmd.Flags().StringSliceVar(&opts.Query.Qualifiers.User, "owner", nil, "Filter on owner")
+
+	return cmd
+}
+
+func codeRun(opts *CodeOptions) error {
+	io := opts.IO
+	if opts.WebMode {
+		url := opts.Searcher.URL(opts.Query)
+		if io.IsStdoutTTY() {
+			fmt.Fprintf(io.ErrOut, "Opening %s in your browser.\n", text.DisplayURL(url))
+		}
+		return opts.Browser.Browse(url)
+	}
+	io.StartProgressIndicator()
+	results, err := opts.Searcher.Code(opts.Query)
+	io.StopProgressIndicator()
+	if err != nil {
+		return err
+	}
+	if len(results.Items) == 0 && opts.Exporter == nil {
+		return cmdutil.NewNoResultsError("no code results matched your search")
+	}
+	if err := io.StartPager(); err == nil {
+		defer io.StopPager()
+	} else {
+		fmt.Fprintf(io.ErrOut, "failed to start pager: %v\n", err)
+	}
+	if opts.Exporter != nil {
+		return opts.Exporter.Write(io, results)
+	}
+
+	return displayResults(io, results, opts.Query.Limit)
+}
+
+func displayResults(io *iostreams.IOStreams, results search.CodeResult, limit int) error {
```

Comment:
In general, `displayResults` ended up being a bit more complicated than I initially expected. I am pretty new to Go, and I am open to suggestions on ways to make this simpler / more idiomatic. 

A few factors at play here:

1. The code search API does not use the `limit` flag as you might expect. The limit seems to apply to the number of files returned, but each file might have multiple 'fragments' returned that match the query, and each 'fragment' can also have multiple matches within it. As discussed [here](https://github.com/cli/cli/issues/6945#issuecomment-1513573499), I used `limit` both when making the request and when rendering the results to stop rendering search results after `limit` rows had been displayed.

2. The code search API (as far as I can tell) does not provide line numbers for the matches. As such, rather than rendering line numbers for each result (as discussed [here](https://github.com/cli/cli/issues/6945#issuecomment-1517999807)), I chose to render each line with a simple counter that tracks with the number of lines printed for each file. I can easily remove this if desired.

3. Some of the complexity of this code is because each line in the 'fragment' can have 0, 1, or many matches within it. We want to skip lines with 0 matches, and print each line with 1 or more matches just once. I have a feeling there is a more elegant solution than what I came up with.

Reply:
I think what you have is solid, and `getHighlightIndices` was pretty clever. I took a quick pass at making it a bit simpler and idiomatic, let me know what you think:
```go
func displayResults(io *iostreams.IOStreams, results search.CodeResult, limit int) error {
	cs := io.ColorScheme()
	tp := tableprinter.New(io)
	displayed := 0
outer:
	for _, code := range results.Items {
		if io.IsStdoutTTY() {
			header := fmt.Sprintf("%s %s", cs.GreenBold(code.Repo.FullName), cs.GreenBold(code.Path))
			tp.AddField(header)
			tp.EndRow()
		}
		for _, match := range code.TextMatches {
			fragments := formatMatch(cs, match)
			for i, fragment := range fragments {
				if io.IsStdoutTTY() {
					tp.AddField(fmt.Sprintf("%s. %s", strconv.Itoa(i+1), fragment))
				} else {
					tp.AddField(fmt.Sprintf("%s %s: %s", code.Repo.FullName, code.Path, fragment))
				}
				tp.EndRow()
				displayed++
				if displayed == limit {
					break outer
				}
			}
		}
	}
	if io.IsStdoutTTY() {
		header := fmt.Sprintf("Showing %d of %d results\n\n", displayed, max(displayed, results.Total))
		fmt.Fprintf(io.Out, "\n%s", header)
	}
	return tp.Render()
}

func formatMatch(cs *iostreams.ColorScheme, tm search.TextMatch) []string {
	shouldHighlight := matchHighlightMask(tm.Fragment, tm.Matches)
	var lines []string
	var found bool
	var b strings.Builder
	for i, c := range tm.Fragment {
		if c == '\n' {
			if found {
				lines = append(lines, b.String())
			}
			found = false
			b.Reset()
			continue
		}
		if shouldHighlight[i] {
			b.WriteString(cs.CyanBold(string(c)))
			found = true
		} else {
			b.WriteRune(c)
		}
	}
	return lines
}

func matchHighlightMask(fragment string, matches []search.Match) []bool {
	m := make([]bool, len(fragment))
	for _, match := range matches {
		start := match.Indices[0]
		stop := match.Indices[1]
		for i := start; i < stop; i++ {
			m[i] = true
		}
	}
	return m
}
```

Reply:
Nice, I haven't utilized labels very often but it is a nice fit here. I incorporated these suggestions, with one small tweak - I had to add back in a variable to keep track of the number of overall displayed matches per item in `results.Items`, so that we can label the results properly as we move through the multiple `TextMatches` possible for each item/file. I haven't found a way a good way to get rid of this variable without dropping the ability to number the matches for each file.